### PR TITLE
bugfix/sliver-app-bar-bottom: Fix constraints

### PR DESCRIPTION
**Pull request by EArminjon**
This PR could solve the following issue :
https://github.com/flutter/flutter/issues/113392

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -1115,14 +1115,19 @@ class _AppBarState extends State<AppBar> {
       appBar = Column(
         mainAxisAlignment: MainAxisAlignment.spaceBetween,
         children: <Widget>[
-          Flexible(
-            child: ConstrainedBox(
-              constraints: BoxConstraints(maxHeight: toolbarHeight),
-              child: appBar,
-            ),
+          if (toolbarHeight == 0)
```
Comment (Piinks) [inline packages/flutter/lib/src/material/app_bar.dart:1118]: This seems really specific to the use case in the issue. What if the toolbar height is 20? or 100? Or 1?

**Review by Piinks — COMMENTED** [review]
Hi @EArminjon this will need a test to verify that it fixes the issue.

**Comment by Piinks** [comment]
This is also breaking a lot of existing tests, can you take a look?

**Comment by EArminjon** [comment]
I apologie for the delay to provide / fix tests.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/app_bar.dart:1118]: Can you provide context on this?

File: packages/flutter/lib/src/material/app_bar.dart [hunk 2]
```diff
@@ -1115,11 +1115,14 @@ class _AppBarState extends State<AppBar> {
...
+            const SizedBox()
+          else
+            Flexible(
```
Comment (Piinks) [inline packages/flutter/lib/src/material/app_bar.dart:1121]: Why would a constrained box be wrapped in a flexible widget? Aren't the constraints supposed to be enforced?

**Review by Piinks — COMMENTED** [review]
The `flexibleSpace` property is supposed to serve for flexible layout. Why do we want to make `bottom` also flexible?

**Comment by EArminjon** [comment]
> The `flexibleSpace` property is supposed to serve for flexible layout. Why do we want to make `bottom` also flexible?

The bottom widget is inside a Column. If in our side use a Flexible or an Expanded as child it will work and the bottom widget will be flexible. This PR is to fix the behavior when toolbarHeight equal zero but took some space.

**Comment by Piinks** [comment]
Hi @EArminjon, apologies this fell off my radar. If you are still interested in this change can you help me understand this last comment:

> This PR is to fix the behavior when toolbarHeight equal zero but took some space.

When would a toolbarHeight be zero but also consume space? That feels like unexpected behavior that could be confusing for developers.

**Comment by EArminjon** [comment]
> Hi @EArminjon, apologies this fell off my radar. If you are still interested in this change can you help me understand this last comment:
> 
> > This PR is to fix the behavior when toolbarHeight equal zero but took some space.
> 
> When would a toolbarHeight be zero but also consume space? That feels like unexpected behavior that could be confusing for developers.

ToolbarHeight is not respected when a developer implement the following code : 

```dart
 SliverAppBar(
            toolbarHeight: 0,
            primary: false,
            expandedHeight: 200,
            pinned: true,
            backgroundColor: Colors.red,
            bottom: PreferredSize(
              preferredSize: const Size.fromHeight(50),
              child: Flexible(
                child: Container(
                  height: 50,
                  color: Colors.green,
                ),
              ),
            ),
          ),
```
This is because AppBar is wrapped inside a Flexible and a Column widget. If expandedHeight = 0 and we use a Flexible in the bottom properties, then appbar and bottom will share the available space which is not logic because toolbarHeight is set to 0.

This PR fix this use case. I agree that it's not a common use case. But in our app, we meet it. Adding a flexible inside bottom properties allow us to do extra animation and things in addition of the flexibleSpace property.

Full example bellow : 
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: Home()));
}

class Home extends StatelessWidget {
  const Home({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: NestedScrollView(
        physics: const ClampingScrollPhysics(),
        headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) => <Widget>[
          const SliverAppBar(
            pinned: true,
            elevation: 0,
            primary: true,
            backgroundColor: Colors.blue,
          ),
          SliverAppBar(
            toolbarHeight: 0,
            primary: false,
            expandedHeight: 200,
            pinned: true,
            elevation: 0,
            backgroundColor: Colors.red,
            bottom: PreferredSize(
              preferredSize: const Size.fromHeight(50),
              child: Flexible( 
                child: Container(
                  height: 50,
                  color: Colors.green,
                ),
              ),
            ),
          ),
        ],
        body: const SizedBox(),
      ),
    );
  }
}
```

**Comment by Piinks** [comment]
Hey @EArminjon, since we have not heard back from you in a while, I am going to close this PR for now to remove it from our review queue.
If you would like to come back and work on it again, feel free to reopen it. Thank you for contributing!

**Pull request closed without merging**